File: packages/updater/src/manifest.ts
export type OtaManifestChannel = {
  name: string;
  stability: string;
  isDefault: boolean;
  latest: string;
  releases: string[];
};

export type ArtifactDigest = {
  size: number;
  sha256: string;
};

export type ReleaseArtifacts = {
  daemon?: ArtifactDigest;
  image_swu?: ArtifactDigest;
  image_zck?: ArtifactDigest;
  image_boot_zck?: ArtifactDigest;
  webapps?: Record<string, ArtifactDigest>;
  daemon_patches?: Record<string, ArtifactDigest>;
};

export type OtaManifestRelease = {
  version: string;
  channel: string;
  yanked: string | null;
  deprecated: boolean;
  builtinWebapps: Record<string, string>;
  artifacts?: ReleaseArtifacts;
};

export type OtaDiscoverManifest = {
  manifestVersion: number;
  updatedAt: string;
  channels: Record<string, OtaManifestChannel>;
  releases: Record<string, OtaManifestRelease>;
};

type RawChannel = {
  name: string;
  stability: string;
  default: boolean;
  latest: string;
  releases: string[];
};

type RawRelease = {
  version: string;
  channel: string;
  yanked: string | null;
  deprecated: boolean;
  builtin_webapps?: Record<string, string>;
  artifacts?: ReleaseArtifacts;
};

type RawManifest = {
  manifest_version: number;
  updated_at: string;
  channels: Record<string, RawChannel>;
  releases: Record<string, RawRelease>;
};

export class ManifestFetchError extends Error {
  constructor(
    message: string,
    public readonly status?: number,
  ) {
    super(message);
    this.name = 'ManifestFetchError';
  }
}

export async function fetchManifest(rootURL: string): Promise<OtaDiscoverManifest> {
  const response = await fetch(`${rootURL}/manifest.json`, { cache: 'no-store' });
  if (!response.ok) {
    throw new ManifestFetchError(`manifest fetch returned HTTP ${response.status}`, response.status);
  }
  const raw = (await response.json()) as RawManifest;
  return {
    manifestVersion: raw.manifest_version,
    updatedAt: raw.updated_at,
    channels: Object.fromEntries(
      Object.entries(raw.channels).map(([id, c]) => [
        id,
        { name: c.name, stability: c.stability, isDefault: c.default, latest: c.latest, releases: c.releases },
      ]),
    ),
    releases: Object.fromEntries(
      Object.entries(raw.releases).map(([version, r]) => [
        version,
        {
          version: r.version,
          channel: r.channel,
          yanked: r.yanked,
          deprecated: r.deprecated,
          builtinWebapps: r.builtin_webapps ?? {},
          artifacts: r.artifacts,
        },
      ]),
    ),
  };
}

export type OtaCompositeVersion = { daemon: string; image: string };

export function parseCompositeVersion(raw: string): OtaCompositeVersion | null {
  const plus = raw.indexOf('+');
  if (plus < 0) return null;
  const daemon = raw.slice(0, plus);
  const suffix = raw.slice(plus + 1);
  const prefix = 'image.';
  if (!suffix.startsWith(prefix)) return null;
  const image = suffix.slice(prefix.length);
  if (daemon.length === 0 || image.length === 0) return null;
  return { daemon, image };
}

export type OtaArtifactURLs = {
  daemonBinary: string;
  imageSwu: string;
  imageZck: string;
  imageBootZck: string;
};

export function imageVariantForChannel(channel: string): string {
  return channel === 'dev' ? 'dev' : 'prod';
}

export function otaArtifactUrls(opts: {
  rootURL: string;
  channel: string;
  daemonVersion: string;
  imageVersion: string;
  imageVariant: string;
}): OtaArtifactURLs {
  const imageName = `bridgething-${opts.imageVariant}-image`;
  const imagesDir = `${opts.rootURL}/images/${opts.channel}/${opts.imageVersion}`;
  return {
    daemonBinary: `${opts.rootURL}/daemon/${opts.channel}/${opts.daemonVersion}/bridgething`,
    imageSwu: `${imagesDir}/${imageName}.swu`,
    imageZck: `${imagesDir}/${imageName}.zck`,
    imageBootZck: `${imagesDir}/${imageName}-boot.zck`,
  };
}

export function daemonPatchUrl(opts: {
  rootURL: string;
  channel: string;
  toVersion: string;
  fromVersion: string;
}): string {
  return `${opts.rootURL}/daemon/${opts.channel}/${opts.toVersion}/patches/from-${opts.fromVersion}.zst`;
}

export function builtinWebappUrl(opts: { rootURL: string; channel: string; name: string; version: string }): string {
  return `${opts.rootURL}/webapps/${opts.channel}/${opts.name}/${opts.version}/${opts.name}.zip`;
}

File: crates/delivery/core/src/ota/manifest.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::bundle::ArtifactDigest;

pub const WAKEWORD_MODEL_FILE: &str = "hey_bridgething.btww";

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all(serialize = "camelCase"))]
pub struct OtaDiscoverManifest {
  pub manifest_version: u32,
  pub updated_at: String,
  pub channels: BTreeMap<String, OtaManifestChannel>,
  pub releases: BTreeMap<String, OtaManifestRelease>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all(serialize = "camelCase"))]
pub struct OtaManifestChannel {
  pub name: String,
  pub stability: String,
  #[serde(rename(deserialize = "default", serialize = "isDefault"))]
  pub is_default: bool,
  pub latest: String,
  pub releases: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all(serialize = "camelCase"))]
pub struct OtaManifestRelease {
  pub version: String,
  pub channel: String,
  pub yanked: Option<String>,
  #[serde(default)]
  pub deprecated: bool,
  #[serde(default)]
  pub builtin_webapps: BTreeMap<String, String>,
  pub wakeword: Option<OtaWakeword>,
  pub artifacts: Option<OtaReleaseArtifacts>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all(serialize = "camelCase"))]
pub struct OtaWakeword {
  /// the daemon version this release's model was trained against; the embedding graph is baked into the daemon.
  pub runtime: String,
  pub model: String,
  #[serde(default)]
  pub model_trained_against: BTreeMap<String, String>,
}

impl OtaWakeword {
  pub fn trained_against(&self, model: &str) -> &str {
    self.model_trained_against.get(model).unwrap_or(&self.runtime)
  }
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all(serialize = "camelCase"))]
pub struct OtaReleaseArtifacts {
  pub daemon: Option<ArtifactDigest>,
  pub daemon_zst: Option<ArtifactDigest>,
  pub image_swu: Option<ArtifactDigest>,
  pub image_zck: Option<ArtifactDigest>,
  pub image_boot_zck: Option<ArtifactDigest>,
  #[serde(default)]
  pub webapps: BTreeMap<String, ArtifactDigest>,
  pub wakeword: Option<OtaWakewordArtifacts>,
  #[serde(default)]
  pub daemon_patches: BTreeMap<String, OtaPatchDigest>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all(serialize = "camelCase"))]
pub struct OtaWakewordArtifacts {
  pub runtime: Option<ArtifactDigest>,
  pub model: Option<ArtifactDigest>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all(serialize = "camelCase"))]
pub struct OtaPatchDigest {
  pub size: u64,
  pub sha256: String,
  pub source_sha256: Option<String>,
}

impl OtaPatchDigest {
  pub fn digest(&self) -> ArtifactDigest {
    ArtifactDigest {
      size: self.size,
      sha256: self.sha256.clone(),
    }
  }
}

pub fn patch_source_matches(declared: Option<&str>, running: Option<&str>) -> bool {
  match (declared, running) {
    (Some(declared), Some(running)) => declared.eq_ignore_ascii_case(running),
    _ => true,
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct OtaCompositeVersion {
  pub daemon: String,
  pub image: String,
}

impl OtaCompositeVersion {
  pub fn parse(raw: &str) -> Option<Self> {
    let (daemon, suffix) = raw.split_once('+')?;
    let image = suffix.strip_prefix("image.")?;
    if daemon.is_empty() || image.is_empty() {
      return None;
    }
    Some(Self {
      daemon: daemon.to_owned(),
      image: image.to_owned(),
    })
  }

  pub fn composite(&self) -> String {
    format!("{}+image.{}", self.daemon, self.image)
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OtaArtifactUrls {
  pub daemon_binary: String,
  pub daemon_binary_zst: String,
  pub image_swu: String,
  pub image_zck: String,
  pub image_boot_zck: String,
}

impl OtaArtifactUrls {
  pub fn build(root_url: &str, channel: &str, daemon_version: &str, image_version: &str, image_variant: &str) -> Self {
    let root = root_url.trim_end_matches('/');
    let image_name = format!("bridgething-{image_variant}-image");

    Self {
      daemon_binary: format!("{root}/daemon/{channel}/{daemon_version}/bridgething"),
      daemon_binary_zst: format!("{root}/daemon/{channel}/{daemon_version}/bridgething.zst"),
      image_swu: format!("{root}/images/{channel}/{image_version}/{image_name}.swu"),
      image_zck: format!("{root}/images/{channel}/{image_version}/{image_name}.zck"),
      image_boot_zck: format!("{root}/images/{channel}/{image_version}/{image_name}-boot.zck"),
    }
  }

  pub fn builtin_webapp(root_url: &str, channel: &str, name: &str, version: &str) -> String {
    let root = root_url.trim_end_matches('/');
    format!("{root}/webapps/{channel}/{name}/{version}/{name}.zip")
  }

  pub fn wakeword_model(root_url: &str, channel: &str, version: &str) -> String {
    let root = root_url.trim_end_matches('/');
    format!("{root}/wakeword/{channel}/model/{version}/{WAKEWORD_MODEL_FILE}")
  }

  pub fn daemon_patch(root_url: &str, channel: &str, to_version: &str, from_version: &str) -> String {
    let root = root_url.trim_end_matches('/');
    format!("{root}/daemon/{channel}/{to_version}/patches/from-{from_version}.zst")
  }
}
